Distributed cache

bergquist opened the pull request:
The goal for this PR is to enable Grafana servers to share information with each other in an HA setup. Previously we used session storage for that in a few cases but since we removed the Session storage for performance reasons we need a new way of sharing information between servers. 

We _could_ implement session data storage with the new auth solution but I felt that this solution would be more generic and usable in more cases. 

By default, Grafana will marshal data and store them in the database using a string key identifier. But it will also support Redis and Memcache to offload the database. While working on this PR I added support for storing all cache data in memory but I removed that since I think this cache should only be used for out of process caching. Making sure it performs the same on one machine and in an HA group. 

I think we should support local and remote caching in Grafana and make it explicit since they are fundamentally different. If not, then I think we should remove local cache and rely on static variables for local caching. 

So far I've only added support for `get`, `set` and `delete` but we could add more operations if we need to. 

One remaining question.. Should we support all three? Only supporting Redis would make it possible to use more features that Redis has to offer. 

closes https://github.com/grafana/grafana/issues/10816

bergquist committed:
cache: initial version of db cache

bergquist committed:
test at interface level instead impl

bergquist committed:
add garbage collector for database cache

bergquist committed:
add support for redis storage

bergquist committed:
add support for memcached

bergquist committed:
fixes typo in redis devenv

bergquist committed:
avoid exposing internal structs and functions

jsferrei commented:
@bergquist I'm curious about this. Is this something being seriously considered? We've been looking for a feature like this so that we can add some safety to our system in the event the main database becomes unavailable for some period of time.

bergquist committed:
extract tests into seperate files

bergquist committed:
adds memory as dist storage alt

bergquist committed:
uses set instead of add for memcache

set always sets the value regardless.

bergquist committed:
avoid exporting test helpers

bergquist committed:
adds config to default settings

bergquist commented:
@jsferrei this will most likley make it into grafana. But I would not consider it a fallback for all datasource read/writes. More like a fast read alternative to using the database. The only reason we allow cache data in a database it to be backwards compatible. 

If you need a HA database setup I think there are better solutions on the database side. Ex http://galeracluster.com/products/

bergquist committed:
test redis and memcached during integration tests

bergquist committed:
removes memory as distcache option

if database caching is to expensive if should
not use distcache in the first place.

bergquist committed:
`memcache` -> `memcached`

https://github.com/memcached/memcached

bergquist committed:
reverts package.json I made during the flight >.>

bergquist committed:
rename put -> set

bergquist committed:
code layouts and comments

bergquist committed:
build steps for cache servers

bergquist committed:
renames key to cache_key

apparently key is a reserved keyword in mysql.
and the error messages doesnt mention that.
can I please have 6h back?

bergquist commented:
GAH! these mysql/postgres build errors are pain

bergquist commented:
Regarding naming... My current thinking is `remoteCache` and rename the current in-memory cache `localCache` making it clear if the data is stored in or out of proc.

bergquist committed:
renames distcache -> remotecache

bergquist committed:
add docs about remote cache settings

bergquist committed:
avoid exposing cache client directly

bergquist committed:
use `Get` instead of `Find`

bergquist commented on .circleci/config.yml:
```diff
@@ -56,6 +56,20 @@ jobs:
             name: postgres integration tests
             command: './scripts/circle-test-postgres.sh'
 
+  cache-server-test:
```
I'm not sure about this. Slowing down the build, even more, feels wrong but at the same time, these things should be tested.

xlson replied:
Should definitely be tested.

bergquist replied:
Perhaps we should add those contains to the backend test task rather than one just for cache servers. What do you think?

xlson replied:
Not sure tbh, maybe test it and see how it affects the time it takes to run backend-test?

bergquist replied:
I'll keep it as it is now and I can experiment more with this later.

bergquist commented on pkg/infra/remotecache/remotecache_test.go:
```diff
@@ -0,0 +1,108 @@
+package remotecache
```
These tests looks a little odd compared to the rest of Grafana but I liked the idea of reusing the same tests for all cache client implementations. 

I liked `"github.com/bmizerany/assert"` (its already vendored) since it makes assert short and precise. 
I think we should use something similar for smaller and tests cases like this and only use convey for scenario testing.

torkelo commented on pkg/infra/remotecache/remotecache.go:
```diff
@@ -0,0 +1,117 @@
+func encodeGob(item *cachedItem) ([]byte, error) {
+	buf := bytes.NewBuffer(nil)
+	err := gob.NewEncoder(buf).Encode(item)
+	return buf.Bytes(), err
```
This feels fragile? if we ever update the data structure of a cached value it will (I think) cause errors here. Is there a way to safeguard that, store in json instead of binary?

torkelo replied:
Or rather the error will be in decodeGob

bergquist replied:
I was considering json but without generics, it would be kinda fugly if we unmarshal structs in the cache package (which I think we should)

`gob` does support encoding/deocding structs that changed as long as they are backward compatible. 

"The source and destination values/types need not correspond exactly. For structs, fields (identified by name) that are in the source but absent from the receiving variable will be ignored. Fields that are in the receiving variable but missing from the transmitted type or value will be ignored in the destination."

https://golang.org/pkg/encoding/gob/#hdr-Types_and_Values

bergquist replied:
@torkelo does that answer your concern?

torkelo replied:
yes

torkelo reviewed (commented):
Great job @bergquist this looks really good! My only concern is what happens if we change data structure on a cached item.

xlson commented on pkg/infra/remotecache/remotecache.go:
```diff
@@ -0,0 +1,117 @@
+// transferred as implementations of interface values need to be registered.
+// Expecting to be used only during initialization, it panics if the mapping
+// between types and names is not a bijection.
+func Register(value interface{}) {
```
Is there any way we could automatically register the structs when we add them to the cache so that we don't have to do it manually?

bergquist replied:
`gob.Register()` can only be called once per type. 
https://github.com/golang/go/blob/master/src/encoding/gob/type.go#L834
https://github.com/golang/go/blob/master/src/encoding/gob/type.go#L820
and its "Expecting to be used only during initialization" 

I like the idea but right now I dont think it's worth pursuing.

xlson commented on pkg/infra/remotecache/remotecache.go:
```diff
@@ -0,0 +1,117 @@
+
+// CacheStorage allows the caller to set, get and delete items in the cache.
+// Cached items are stored as byte arrays and marshalled using "encoding/gob"
+// so any struct added to the cache needs to be registred with `distcache.Register`
```
Should say `remotecache` instead of `distcache` right?

xlson commented on pkg/infra/remotecache/database_storage.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	var expiresAtEpoch int64
+	if expire != 0 {
```
Any reason in particular that we don't have a cache_max_duration? If we were just saving things in memcached it wouldn't really matter, but with redis and mysql we might end up with some data staying around forever (which probably shouldn't be in a cache to start with).

bergquist replied:
Good catch. I'll add support for that.

xlson commented on pkg/infra/remotecache/remotecache_test.go:
```diff
@@ -0,0 +1,108 @@
+	err := client.Set("key1", cacheableStruct, 0)
+	assert.Equal(t, err, nil)
+
+	// should not be able to read that value since its expired
```
remove not

xlson reviewed (changes requested):
Some minor things. Great job!

bergquist commented on pkg/infra/remotecache/remotecache.go:
```diff
@@ -0,0 +1,117 @@
+	return ctx.Err()
+}
+
+func createClient(opts *setting.RemoteCacheOptions, sqlstore *sqlstore.SqlStore) CacheStorage {
```
This func should not default to database. it should return error if its miss configured

bergquist committed:
return error if cache type is invalid

bergquist committed:
dont allow inifinite expiration

bergquist committed:
updates old distcache names

bergquist committed:
use constants for cache type

bergquist commented:
Fixed all remarks. Please review the last 4 commits.

xlson reviewed (approved):
Looks great!
